### DISCUSSION TITLE
lack of good autocompletion support for json5

### DISCUSSION THREAD

**Discussion by reegnz** [renovatebot/renovate#24433] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

Never worked.

### Wanted end result.

I want to edit the `renovate.json5` config files easily in vscode and neovim (and any other editor supporting LSP-s).
Easy means having proper syntax highlighting, autoformatting, and schema-based autocompletion.

When using plain json config format all of the above just works.

When I use json5 with just the comments feature and don't use any of it's features it works fine for me in neovim (if I force the filetype to be jsonc instead of json5).

There is only https://marketplace.visualstudio.com/items?itemName=mrmlnc.vscode-json5 for syntax, but neither autoformatting nor autocompletion is supported.
I guess autoformatting can be solved with prettier but autocompletion is still a pretty big hit.

Any suggestion on how to improve json5 IDE support for renovate configs?
I feel like after I've switched to json5 it's an objectively worse experience when losing IDE autoformatting and IDE autocompletion.

I'd even propose introducing support for the `renovate.jsonc` extension just so it's less of a hassle for people to integrate with IDE-s when the only thing they want is comments and don't care about the rest of the json5 features.

### What you tried so far.

Forcing the IDE to recognize `renovate.json5` as `jsonc`:

In vscode I use this config to force `jsonc` filetype for our devs:
```jsonc
{
    "files.associations": {
        "renovate.json5": "jsonc" // force jsonc for lack of good json5 IDE support
    }
}
```

For neovim I force it with this comment in the first line:
```
// vim: ft=jsonc :
```

**Comment by HonkingGoose** [comment]
This issue seems related:

- https://github.com/renovatebot/renovate/issues/24554

I'll let the maintainers respond to the `.json`/`.json5`/`.jsonc` autocomplete for Renovate stuff.

**Reply by rarkins** [reply]
I don't think this is Renovate's problem to solve. We don't force people to use JSON5, if they do then if IDEs don't support it great out of the box then we are not going to change the IDEs

**Reply by reegnz** [reply]
I think my point was that supporting jsonc as well would be low overhead, and can be parsed as JSON5, and the jsonc extension would be picked up properly by IDE-s.

IMHO having your config formats working nice with IDE-s is 100% Renovate's problem as renovate chose to support json5 as a config format.

Right now your choice is good IDE support but you cannot easily document your config with comments, vs. terrible IDE support, but you have comments to document your config.
Middle ground would be supporting jsonc format.

**Reply by HonkingGoose** [reply]
> IMHO having your config formats working nice with IDE-s is 100% Renovate's problem as Renovate chose to support JSON5 as a config format.

That depends on whether the Renovate maintainers want to support _all_ file formats equally, or if they say: "for the best experience use the `renovate.json` format". The Renovate docs should explain what's the best format of course. 😉 

I guess only plain `.json` benefits from [VSCode's JSON schema support](https://code.visualstudio.com/Docs/languages/json#_json-with-comments)?

Here's the `renovate.json` schema link:

https://github.com/renovatebot/renovate/blob/a6c9075f76fc747070873719a34530818ed16f73/renovate.json#L2

I found the [JSON Schema organization: JSON Core Schema](https://json-schema.org/draft/2020-12/json-schema-core.html), maybe that explains what file formats are supported with the schema?

**Reply by rarkins** [reply]
The entire industry has a problem if:
- JSON for config files is insufficient due to lack of commenting, and
- IDEs don't natively support extended JSON

And we can't fix the entire industry.

> 100% Renovate's problem as renovate chose to support json5 as a config format.

I don't agree at all on this. If it were truly a _problem we created_ then that would imply that one solution is _removing_ JSON5 support. But that would be counterproductive because (a) we don't make people use JSON5, and (b) some users would be very sad to lose it.

**Reply by reegnz** [reply]
@HonkingGoose not just plain json benefits from json-schema, I also get autocomplete support for jsonc files if the schema is defined (both in neovim with json-lsp and vscode). json-schema even works for yaml but that's a different language server / vscode plugin.
